File: benchmarks/poly/jsondeserialize/jsondeserialize.js
function main() {
    const _t0 = process.hrtime.bigint();
    const n = 400000;
    let checksum = 0;
    let totalNameLen = 0;

    for (let i = 0; i < n; i++) {
        const line = '{"id":' + i + ',"name":"user' + i + '","score":' + (i % 100) + "}";

        const idStart = line.indexOf(":") + 1;
        const idEnd = line.indexOf(",");
        const idVal = parseInt(line.slice(idStart, idEnd), 10);

        const namePrefix = '"name":"';
        const nameStart = line.indexOf(namePrefix) + namePrefix.length;
        const nameEnd = line.indexOf('"', nameStart);
        const nameVal = line.slice(nameStart, nameEnd);

        const scoreStart = line.lastIndexOf(":") + 1;
        const scoreEnd = line.lastIndexOf("}");
        const scoreVal = parseInt(line.slice(scoreStart, scoreEnd), 10);

        checksum = (checksum * 31 + idVal + scoreVal) % 1000000007;
        totalNameLen += nameVal.length;
    }

    console.log(checksum);
    console.log(totalNameLen);
    console.error(`bench_ns=${process.hrtime.bigint() - _t0}`);
}

main();

File: benchmarks/poly/jsondeserialize/jsondeserialize.rs
fn main() {
    let t0 = std::time::Instant::now();
    let n = 400000;
    let mut checksum: i64 = 0;
    let mut total_name_len: i64 = 0;

    for i in 0..n {
        let line = format!(r#"{{"id":{},"name":"user{}","score":{}}}"#, i, i, i % 100);

        let id_start = line.find(':').unwrap() + 1;
        let id_end = line.find(',').unwrap();
        let id_val: i64 = line[id_start..id_end].parse().unwrap();

        let name_prefix = r#""name":""#;
        let name_start = line.find(name_prefix).unwrap() + name_prefix.len();
        let name_end = name_start + line[name_start..].find('"').unwrap();
        let name_val = &line[name_start..name_end];

        let score_start = line.rfind(':').unwrap() + 1;
        let score_end = line.rfind('}').unwrap();
        let score_val: i64 = line[score_start..score_end].parse().unwrap();

        checksum = (checksum * 31 + id_val + score_val) % 1000000007;
        total_name_len += name_val.len() as i64;
    }

    println!("{}", checksum);
    println!("{}", total_name_len);
    eprintln!("bench_ns={}", t0.elapsed().as_nanos());
}
